Add some type signatures

pjonsson opened the pull request.

omad commented:
@pjonsson 
Is there any benefit having `-> None` specified for pytest test functions? They shouldn't ever be called directly. I don't know if it exists, but to me a linting rule that test functions aren't returning anything would make more sense. And failing that, nothing is going to break even if someone does mistakenly return something.

It just seems like unnecessary noise... But happy to reconsider if there is a benefit.

omad reviewed (changes requested):
Above question about defining None return types on test functions.

Kirill888 commented:
> Is there any benefit having `-> None` specified for pytest test functions?

for many of them it enables `mypy` checking, as they had no annotation at all prior to the change. Question is: do we need mypy checking in tests at all? It might be beneficial in simpler tests, but when you start using mocks and such it only gets in the way.

Will this PR make change management easier going forward or will it only introduce extra hurdles and random failures of the type: "almost nothing changed in the codebase, but mypy now complains about some test nobody touched in a year"?

pjonsson commented:
As Kirill says, this will enable type checking on a lot of these functions. My experience with modern type checkers is that they are often finding real bugs early and with less effort than it would have taken to debug the symptom (even for someone who knows the code base). I'm not saying type checkers have a zero cost, it obviously takes time to add annotations to the code base, and those annotations sometimes uncover real bugs that you then have to spend time on fixing. There's also a cost of occasionally having to fight the type checker over some piece of code that is correct but the type checker refuses to accept, which I believe is what Kirill has in mind with the mocks.

There are also some secondary effects of having the None return values present, I as a drive-by contributor knows that the function was not meant to return anything, and so does my IDE and any static analysis tools applied to the code base. For pytest test functions, I'm aware that they shouldn't return anything, but it's nice to have that mechanically checked.

I'm not suggesting the entire code base should have type annotations and no ignores, add an ignore for situations like when the mock is correct but the type checker refuses to accept that.

pjonsson commented:
> Will this PR make change management easier going forward or will it only introduce extra hurdles and random failures of the type: "almost nothing changed in the codebase, but mypy now complains about some test nobody touched in a year"?

I can share some anecdotal evidence concerning a similar situation. The difference is that I switched to using Pyright about a year ago, and it's obviously for a different code base (whose code coverage has been 83-88% from the tests).

During that year, I've had two Pyright version updates cause the CI to fail. The first one was because Pyright had updated the TypeShed files distributed in the release, so improved type annotations of third-party modules found bugs in the code. Things happened to work the way things were written, but I'm fairly sure that having Pyright fail avoided hitting a land mine in the future, so the afternoon spent on fixing that was well worth its time. The second time it failed was because Pyright had improved something (or enabled some setting by default), but fortunately the detected problem was only present in the tests, and it took me 30 minutes to fix.

I started from a mostly type annotated code base that type checked with MyPy, and switching to Pyright "forced" me to add a couple of ignores that I didn't think much about. A couple of weeks later I spent 2 days on tracking down and cleaning up after some mysterious bug that in the end turned out to be caused by the code that I had added an ignore for when switching to Pyright. In hindsight, I should obviously have fixed the code right away because in total I've spent 2 days on a single bug and less than a day on everything flagged by the type checker, but the code hadn't been touched for years and I wasn't sure about any domino-effects from changing it.

It's obviously not going to be a 1:1 experience, but the code base does use shapely/rasterio/SQLAlchemy and my impression is that the opendatacube code base is of comparable quality, so my guess is that most of the time you'll be fine when updating MyPy versions, and occasionally it will find a real bug. It might be too much work to fix a bug though, but for those times it's easy to add an ignore.

pjonsson commented:
I have the impression that everybody agrees that it is a good idea to have type signatures and type checking of the library code, so the question is about the cost versus benefit of having them for the tests. Please let me know if you feel I'm misrepresenting what you wrote Kirill.

I completely agree with the list of possible causes for the type checker suddenly failing on code that previously type checked. The events described can definitely happen, so we end up with a risk assessment, and I feel it's beneficial to consider the specifics for this repository when doing that. If the type checker is already used for the library code, the risk of having to bump version ranges of dependencies for compatibility reasons is already present, so having type checked tests do not increase that risk.

We do have some statistics of additional type checking of tests in this repository from this PR though: the diffstat is +54-50. There are 3 lines subtracted in the category that I'll label as "other", and 47 lines subtracted that add new type annotations. The library code gets 6 of those type annotations, and the tests get the 41 remaining ones. Breaking down those 41 lines:
* 38 tests were unaffected. Some of the tests already had type annotations for other parts of the function, and the rest passed the type checker despite never having been checked before.
* 2 NaN tests had to have their assertions split up to convey to the type checker that they were not passing None to the tested function, so they were responsible for 2 out of the 3 lines of the "other" category I just mentioned.
* The remaining line from the "other" category seems like a genuine bug in a previously unchecked test (`test_driver_load`), which I've fixed.

> [..] I don't think that's a strong claim, most methods are "self-tested" as they are being called by other type annotated methods within non-test code.

Sure, finding 1 bug with 41 attempts is a 2.4% hit rate, and that is pretty bad. Time wise (since I had my docker container up and running) I spent a total of 10 minutes to find this bug and fix it along with changing the two NaN tests so they passed the type checker.

omad reviewed (approved):
Thanks for the great discussion @Kirill888 and @pjonsson .

While I still don't think there's much benefit to having type checked tests when the tests are all running correctly, I don't expect there to be much downside either.

I guess we'll find out later, but any incoming changes to libraries/mypy/other type checkers is going to be a problem for library code at least as much as for tests.
